tests/functional/a/alternative/alternative_union_syntax_error.py

```diff
@@ -87,3 +87,44 @@ class CustomDataClass3:
 @dataclasses.dataclass
 class CustomDataClass4:
     my_var: int | str  # [unsupported-binary-operation]
+
+# Not an error if the metaclass implements __or__
+
+class ForwardMetaclass(type):
+    def __or__(cls, other):
+        return True
+
+class ReverseMetaclass(type):
+    def __ror__(cls, other):
+        return True
+
+class WithForward(metaclass=ForwardMetaclass):
+    pass
+
+class WithReverse(metaclass=ReverseMetaclass):
+    pass
+
+class DefaultMetaclass:
+    pass
+
+class_list = [WithForward | DefaultMetaclass]
+class_list_reversed_invalid = [WithReverse | DefaultMetaclass] # [unsupported-binary-operation]
+class_list_reversed_valid = [DefaultMetaclass | WithReverse]
+
+
+# Pathological cases
+class HorribleMetaclass(type):
+    __or__ = lambda x: x
+
+class WithHorrible(metaclass=HorribleMetaclass):
+    pass
+
+class_list = [WithHorrible | DefaultMetaclass]
+
+class SecondLevelMetaclass(ForwardMetaclass):
+    pass
+
+class WithSecondLevel(metaclass=SecondLevelMetaclass):
+    pass
+
+class_list_with_second_level = [WithSecondLevel | DefaultMetaclass]
```

Comment:
This doesn't seem to work at the moment. I only spend a few minutes investigating it, but couldn't  a quick fix. Maybe it's better to remove that test case here and open a followup issue for it to be addressed later. At least that way this PR could be merged.

Reply:
I was hoping to get to the bottom of it, because I didn't want to commit a half-working fix. But thinking about it, this failure case isn't a new false positive so I guess this fix will be an improvement overall. I'll pull this out and open another PR for it.